src/function/gds/output_writer.cpp

```diff
@@ -104,99 +107,139 @@ void PathsOutputWriter::write(processor::FactorizedTable& fTable, nodeID_t dstNo
+    }
+    return result;
+}
+
+// This code checks if we should switch from backtracking to forward-tracking, i.e.,
+// moving forward in the DFS logic to find paths. We switch from backtracking if:
+bool PathsOutputWriter::canNextReplacePathTop(ParentList* next,
+    const std::vector<ParentList*>& path) const {
+    if (next == nullptr) {
+        return false;
+    }
+    auto nextIter = next->getIter();
+    // (1) if this is the first element in the stack (curPath.size() == 1), i.e., we
+    // are enumerating the parents of the destination, then we should switch to
+    // forward-tracking if the next parent has visited the destination at a length
+    // that's greater than or equal to the lower bound of the recursive join. Otherwise,
+    // we'll enumerate paths that are smaller than the lower; OR
+    if (path.size() == 1 && nextIter >= info.lowerBound) {
         return true;
     }
-    // Path node predicate should only be applied to intermediate node. So we exclude dst node.
-    for (auto i = 0u; i < path.size() - 1; ++i) {
-        if (!info.pathNodeMask->valid(path[i]->getNodeID())) {
-            return false;
-        }
+    // (2) if this is not the first element in the stack, i.e., then we should switch
+    // to forward tracking only if the next parent of the top node in the stack has the
+    // same iter value as the current parent. That's because the levels/iter need to
+    // decrease by 1 each time we add a new node in the stack.
+    if (nextIter == getTop(path)->getIter()) {
+        return true;
     }
-    return true;
+    return false;
+}
+
+bool PathsOutputWriter::checkNewPathElement(const std::vector<ParentList*>& path,
+    ParentList* element) const {
+    return checkPathNodeMask(element) && checkSemantic(path, element);
 }
 
-// TODO(Xiyang): apply semantic check during DFS backtracking.
-bool PathsOutputWriter::checkSemantic(const std::vector<ParentList*>& path) const {
+bool PathsOutputWriter::checkPathNodeMask(ParentList* element) const {
+    if (!info.hasNodeMask() || element->getIter() == 1) {
+        return true;
+    }
+    return info.pathNodeMask->valid(element->getNodeID());
+}
+
+bool PathsOutputWriter::checkSemantic(const std::vector<ParentList*>& path,
+    ParentList* candidate) const {
     switch (info.semantic) {
     case PathSemantic::WALK:
         return true;
     case PathSemantic::TRAIL:
-        return isTrail(path);
+        return isTrail(path, candidate);
     case PathSemantic::ACYCLIC:
-        return isAcyclic(path);
+        return isAcyclic(path, candidate);
     default:
         KU_UNREACHABLE;
     }
 }
 
-bool PathsOutputWriter::isTrail(const std::vector<ParentList*>& path) const {
-    rel_id_set_t set;
-    for (auto i = 0u; i < path.size(); ++i) {
-        auto edgeID = path[i]->getEdgeID();
-        if (set.contains(edgeID)) {
+bool PathsOutputWriter::isTrail(const std::vector<ParentList*>& path, ParentList* candidate) const {
+    for (auto& element : path) {
```

Comment:
I don't think we should check it this way. You are implementing a quadratic algorithm here, looping through the entire path each time you add a new element. Instead, I would maintain an unordered map of nodes (for acyclic) or edges (for trail) along with the currentPath as you construct path and add/remove elements from it each time I add/remove a new node/edge from the path. Then you can do these checks with 1 map call (using a node or edge). We also need a fast path that does not do any map operations for the walk semantics.

You need to benchmark and compare both these implementations to see that using a map is faster than your quadratic/for-loop-based implementation, and then that using map is also faster than the post-enumeration checks (which is what's done prior to this PR). 

This requires some serious work. This won't be quick unfortunately, but needs to be done.

Reply:
Consider the size of path is always small (less than 30). I doubt using a map will be noticeable faster than for loop. But I have opened an issue to track this. Will take a look in the future.